### PR TITLE
Support ValueTask as [OneWay] Methods Return Type

### PR DESCRIPTION
**Pull request by ThiagoT1**

### CODE REVIEW

File: src/Orleans.CodeGeneration/GrainReferenceGenerator.cs [hunk 1]
```diff
@@ -209,15 +209,21 @@ private static MemberDeclarationSyntax[] GenerateInvokeMethods(Type grainType, A
...
                     {
-                        if (method.ReturnType != typeof(Task))
+                        if (method.ReturnType == typeof(Task))
+                        {
+                            var done = typeof(Task).GetNameSyntax(true).Member((object _) => Task.CompletedTask);
+                            body.Add(SF.ReturnStatement(done));
+                        }
+                        else if (method.ReturnType == typeof(ValueTask))
+                        {
+                            body.Add(SF.ReturnStatement(SF.ParseExpression("default(System.Threading.Tasks.ValueTask)")));
```
Comment (ThiagoT1) [inline src/Orleans.CodeGeneration/GrainReferenceGenerator.cs:219]: I know this is not pretty. 
Maybe you guys can guide me on how to fix this?

File: src/Orleans.CodeGenerator/Generators/GrainReferenceGenerator.cs [hunk 2]
```diff
@@ -195,15 +196,23 @@ private MemberDeclarationSyntax[] GenerateInvokeMethods(GrainInterfaceDescriptio
...
                     {
-                        if (!SymbolEqualityComparer.Default.Equals(wellKnownTypes.Task, method.ReturnType))
+                        if (SymbolEqualityComparer.Default.Equals(wellKnownTypes.Task, method.ReturnType))
+                        {
+                            var done = wellKnownTypes.Task.ToNameSyntax().Member((object _) => Task.CompletedTask);
+                            body.Add(ReturnStatement(done));
+                        }
+                        else if (SymbolEqualityComparer.Default.Equals(wellKnownTypes.ValueTaskNamed, method.ReturnType))
+                        {
+                            body.Add(ReturnStatement(ParseExpression("default(System.Threading.Tasks.ValueTask)")));
```
Comment (ThiagoT1) [inline src/Orleans.CodeGenerator/Generators/GrainReferenceGenerator.cs:206]: I know this is not pretty. 
Maybe you guys can guide me on how to fix this?

File: src/Orleans.CodeGeneration/GrainReferenceGenerator.cs (hunk 1 above)
Comment (ReubenBond) [reply src/Orleans.CodeGeneration/GrainReferenceGenerator.cs:219]: Would this work?

```c#
body.Add(SF.ReturnStatement(SF.DefaultExpression(typeof(ValueTask).GetTypeSyntax())))
```

Comment (ThiagoT1) [reply src/Orleans.CodeGeneration/GrainReferenceGenerator.cs:219]: Tests are passing, yes.
Also, checked generated code. It's correct.

Comment (ThiagoT1) [reply src/Orleans.CodeGeneration/GrainReferenceGenerator.cs:219]: example:
```csharp

global::System.Threading.Tasks.Task global::UnitTests.GrainInterfaces.IOneWayGrain.Notify(global::UnitTests.GrainInterfaces.ISimpleGrainObserver observer0)
{
	global::Orleans.CodeGeneration.GrainFactoryBase.CheckGrainObserverParamInternal(observer0);
	base.InvokeOneWayMethod(unchecked((int)0x92B0AD04), new object[]{observer0 is global::Orleans.Grain ? observer0.AsReference<global::UnitTests.GrainInterfaces.ISimpleGrainObserver>() : observer0});
	return global::System.Threading.Tasks.Task.CompletedTask;
}

global::System.Threading.Tasks.ValueTask global::UnitTests.GrainInterfaces.IOneWayGrain.NotifyValueTask(global::UnitTests.GrainInterfaces.ISimpleGrainObserver observer0)
{
	global::Orleans.CodeGeneration.GrainFactoryBase.CheckGrainObserverParamInternal(observer0);
	base.InvokeOneWayMethod((int)0x7F9DCB82, new object[]{observer0 is global::Orleans.Grain ? observer0.AsReference<global::UnitTests.GrainInterfaces.ISimpleGrainObserver>() : observer0});
	return default(System.Threading.Tasks.ValueTask);
}


```

**Comment by ThiagoT1** [comment]
Need guidance on how to fix Azure DevOps - BVT:


CodeGenerator.Tests.RoslynTypeNameFormatterTests.TypeCodesMatch
System.InvalidOperationException : Unable to find type with metadata name "System.Threading.Tasks.ValueTask".

File: src/Orleans.CodeGeneration/GrainReferenceGenerator.cs (hunk 1 above)
Comment (ReubenBond) [reply src/Orleans.CodeGeneration/GrainReferenceGenerator.cs:219]: Can you push the changed to fix this here and below?

Comment (ThiagoT1) [reply src/Orleans.CodeGeneration/GrainReferenceGenerator.cs]: Done. 
Generated code is like this now:

```csharp

global::System.Threading.Tasks.Task global::UnitTests.GrainInterfaces.IOneWayGrain.Notify(global::UnitTests.GrainInterfaces.ISimpleGrainObserver observer0)
{
	global::Orleans.CodeGeneration.GrainFactoryBase.CheckGrainObserverParamInternal(observer0);
	base.InvokeOneWayMethod(unchecked((int)0x92B0AD04), new object[]{observer0 is global::Orleans.Grain ? observer0.AsReference<global::UnitTests.GrainInterfaces.ISimpleGrainObserver>() : observer0});
	return global::System.Threading.Tasks.Task.CompletedTask;
}

global::System.Threading.Tasks.ValueTask global::UnitTests.GrainInterfaces.IOneWayGrain.NotifyValueTask(global::UnitTests.GrainInterfaces.ISimpleGrainObserver observer0)
{
	global::Orleans.CodeGeneration.GrainFactoryBase.CheckGrainObserverParamInternal(observer0);
	base.InvokeOneWayMethod((int)0x7F9DCB82, new object[]{observer0 is global::Orleans.Grain ? observer0.AsReference<global::UnitTests.GrainInterfaces.ISimpleGrainObserver>() : observer0});
	return default;
}


```

**Comment by ReubenBond** [comment]
Thank you, @ThiagoT1!

File: src/Orleans.CodeGenerator/WellKnownTypes.cs [hunk 3]
```diff
@@ -63,6 +63,7 @@ public WellKnownTypes(Compilation compilation)
             SerializerFeature = Type("Orleans.Serialization.SerializerFeature");
             String = compilation.GetSpecialType(SpecialType.System_String);
             Task = Type("System.Threading.Tasks.Task");
+            ValueTaskNamed = Type("System.Threading.Tasks.ValueTask");
```
Comment (pentp) [inline src/Orleans.CodeGenerator/WellKnownTypes.cs:66]: Why this duplicate entry here?
There is a perfectly good entry two lines below already that doesn't break `netstandard 2.0` compatibility.

**Pull request merged**